[MM-47929] Switch to using mm-desktop:// protocol for navigating through the app

devinbinnie opened the pull request:
#### Summary
In order to effectively separate any origin-based storage system with the integration, I've added a special new protocol to the app called "mm-desktop" that it used as a substitute for the local `file:///` URL we've been using.

It will prefix the files with `mm-desktop://<server-host-name>`, so that Chrome knows that anything loaded under the URL should use it as its origin.

In order to facilitate this, we've had to do some CORS magic in the main process as well, effectively adding a proxy that the renderer process will hit and verify on behalf of the server that any cross-origin requests between the `mm-desktop` origin and the server are okay.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-47929

```release-note
NONE
```

jupenur commented on src/main/views/MattermostView.ts:
```diff
@@ -96,7 +93,64 @@ export class MattermostView extends EventEmitter {
+    }
+
+    private extractCORSHeaders = (details: OnBeforeSendHeadersListenerDetails) => {
+        if (!details.url.match(new RegExp(`${this.tab.server.url.origin}/(.+)`))) {
```
The RegExp needs to be anchored to the beginning of the string, otherwise a request to a crafted URL like `https://example.com/?https://community.mattermost.com/` matches. Dots and square brackets (valid in IPv6 URLs) in the origin also need to be escaped. This applies everywhere where this type of check is being used.

Also personally I would prefer `RegExp.prototype.test` over `String.prototype.match` where a boolean is expected.

Really this shouldn't be using RegExps at all, though.

jupenur commented on src/main/app/initialize.ts:
```diff
@@ -271,6 +281,10 @@ function initializeAfterAppReady() {
     const defaultSession = session.defaultSession;
 
+    defaultSession.protocol.registerFileProtocol('mm-desktop', (request, callback) => {
+        callback(request.url.replace(/mm-desktop:\/\/([A-Za-z0-9.]+)\//, '').replace(/#(.+)/, ''));
```
Missing anchoring here as well.

jupenur reviewed (changes requested):
Most of the regexps seem to be missing anchoring, and really you should be avoiding regexps in general.

For a proper review I need to test a bunch of things, so can you make sure this builds properly on CI first? My local dev environment is WSL and it can't build the desktop app.

devinbinnie commented:
> Most of the regexps seem to be missing anchoring, and really you should be avoiding regexps in general.
> 
> For a proper review I need to test a bunch of things, so can you make sure this builds properly on CI first? My local dev environment is WSL and it can't build the desktop app.

Thanks for the review @jupenur. I've updated the regex's with anchors and fixed the tests.

As for not using regexes at all, can you think of another way to match the URL without them? A simple `startsWith` would work in a few cases but we'd be adding a lot more cases and doesn't really solve the issue. That's all I can think of at the moment.

mattermod commented:
Building app in separate branch.

mattermod commented:
Successfully building:  
https://circleci.com/gh/mattermost/desktop/26138  
https://circleci.com/gh/mattermost/desktop/26137  
https://circleci.com/gh/mattermost/desktop/26136

jupenur commented:
> As for not using regexes at all, can you think of another way to match the URL without them?

Most of these checks will be looking at two things: the origin and some arbitrary features of the path. Using a regex is a neat way of tackling both at once, but really they're two different things. The origin actually has relevance for security, because if we mix up two different origins, headers get attached to the wrong request, cookies end up in the wrong place, or the wrong URL gets rewritten. If we mix up the path, nothing dramatic will happen; the app might not behave entirely as intended, but it also won't be sending anyone's credentials to an attacker's server.

For that reason my recommendation would be to actually check the two separately. For the origin you should ideally be using a URL parser, or if for some reason one isn't available, the simplest possible `indexOf` / `startsWith` style check. For the path (and any other stuff) using a regexp is fine, although I think in most cases something else will be more efficient: for these types of blocking network event handlers, performance can actually be fairly important and compiling and evaluating multiple regexps every time can add up.

m1lt0n reviewed (approved):
Looks good to me (code-wise), the important review here is the one from Juho, to see if there are security concerns with the new solution as well.

devinbinnie commented:
/update-branch

devinbinnie commented:
/update-branch

devinbinnie commented:
@tboulis @m1lt0n @jupenur Did a bit of a re-work here, now we check for the host and protocol separately from the rest of the URL before parsing and rewriting. Hopefully that makes sense (would have used `origin` but apparently for non-standard URL schemes it's always null 🙄)

Asking for re-reviews across the board to make sure everyone's on board here.

devinbinnie commented on src/main/webRequest/webRequestManager.ts:
```diff
@@ -45,12 +47,25 @@ export class WebRequestManager {
 
-            if (!details.url.match(regex)) {
+            const parsedURL = url.parseURL(details.url);
+            if (!parsedURL) {
+                return {};
+            }
+
+            if (parsedURL.host !== host) {
+                return {};
+            }
+
+            if (parsedURL.protocol !== protocol) {
+                return {};
+            }
+
+            if (!parsedURL.pathname.match(pathnameRegex)) {
```
I just separated them for clarity's sake

devinbinnie commented on src/main/views/MattermostView.ts:
```diff
@@ -183,28 +177,65 @@ export class MattermostView extends EventEmitter {
-        if (url.endsWith('/')) {
-            url = url.slice(0, url.length - 1);
+    private extractCORSHeaders = (details: OnBeforeSendHeadersListenerDetails) => {
+        if (!details.url.startsWith(this.tab.server.url.origin)) {
```
Probably, but if the server URL is undefined there's likely a bigger problem going on.

jupenur commented:
> would have used `origin` but apparently for non-standard URL schemes it's always null 🙄

This is is accurate, but you're explicitly declaring the scheme as standard so origin should work. I just tested it with the build from CircleCI: 

```
> new URL("mm-desktop://potato/path").origin
'mm-desktop://potato'
> new URL("xx-desktop://potato/path").origin
'null'
```

jupenur reviewed (commented):
None of the `mm-desktop:` URLs are resolving for me:

```
(node:17776) electron: Failed to load URL: mm-desktop://community.mattermost.com/playbooks/ with error: ERR_FILE_NOT_FOUND
```
The path passed to the `registerFileProtocol` callback looks correct though: `C:/Users/Juho%20Nurminen/Downloads/mattermost-desktop-5.3.0-develop.1-win64/resources/app.asar/renderer/mattermost.html`

devinbinnie commented:
> > would have used `origin` but apparently for non-standard URL schemes it's always null 🙄
> 
> This is is accurate, but you're explicitly declaring the scheme as standard so origin should work. I just tested it with the build from CircleCI:
> 
> ```
> > new URL("mm-desktop://potato/path").origin
> 'mm-desktop://potato'
> > new URL("xx-desktop://potato/path").origin
> 'null'
> ```
> 
> Edit: nvm, looks like it only works correctly in the renderer context ¯\_(ツ)_/¯

Oh, Electron you silly framework XD

devinbinnie commented:
> None of the `mm-desktop:` URLs are resolving for me:
> 
> ```
> (node:17776) electron: Failed to load URL: mm-desktop://community.mattermost.com/playbooks/ with error: ERR_FILE_NOT_FOUND
> ```
> 
> The path passed to the `registerFileProtocol` callback looks correct though: `C:/Users/Juho%20Nurminen/Downloads/mattermost-desktop-5.3.0-develop.1-win64/resources/app.asar/renderer/mattermost.html`

Ahh, must be something wrong with the release build version. I'll have a look.

mattermost-build commented:
Building app in separate branch.

mattermost-build commented:
Successfully building:  
https://circleci.com/gh/mattermost/desktop/26989  
https://circleci.com/gh/mattermost/desktop/26988  
https://circleci.com/gh/mattermost/desktop/26987

mattermost-build commented:
Artifact links:  
https://output.circle-artifacts.com/output/job/87a325f5-1cb3-432e-8e62-06bf6a486401/artifacts/0/tmp/artifacts/mattermost-desktop-5.3.0-develop.1-linux-arm64.tar.gz  
https://output.circle-artifacts.com/output/job/87a325f5-1cb3-432e-8e62-06bf6a486401/artifacts/0/tmp/artifacts/mattermost-desktop-5.3.0-develop.1-linux-x64.tar.gz  
https://output.circle-artifacts.com/output/job/63de13ff-0b21-4bda-91e4-ac5429810b39/artifacts/0/tmp/artifacts/mattermost-desktop-5.3.0-develop.1-mac-m1.dmg  
https://output.circle-artifacts.com/output/job/63de13ff-0b21-4bda-91e4-ac5429810b39/artifacts/0/tmp/artifacts/mattermost-desktop-5.3.0-develop.1-mac-x64.dmg  
https://output.circle-artifacts.com/output/job/506a75eb-2a92-4cd5-b468-f9d0234e4839/artifacts/0/tmp/artifacts/mattermost-desktop-5.3.0-develop.1-win-arm64.zip  
https://output.circle-artifacts.com/output/job/506a75eb-2a92-4cd5-b468-f9d0234e4839/artifacts/0/tmp/artifacts/mattermost-desktop-5.3.0-develop.1-win-ia32.zip  
https://output.circle-artifacts.com/output/job/506a75eb-2a92-4cd5-b468-f9d0234e4839/artifacts/0/tmp/artifacts/mattermost-desktop-5.3.0-develop.1-win-x64.zip

mattermost-build commented:
Building app in separate branch.

mattermost-build commented:
Successfully building:  
https://circleci.com/gh/mattermost/desktop/27007  
https://circleci.com/gh/mattermost/desktop/27006  
https://circleci.com/gh/mattermost/desktop/27005

mattermost-build commented:
Artifact links:  
https://output.circle-artifacts.com/output/job/ed8d55f2-f56b-4ed0-83ff-ce79023240a6/artifacts/0/tmp/artifacts/mattermost-desktop-5.3.0-develop.1-mac-m1.dmg  
https://output.circle-artifacts.com/output/job/ed8d55f2-f56b-4ed0-83ff-ce79023240a6/artifacts/0/tmp/artifacts/mattermost-desktop-5.3.0-develop.1-mac-x64.dmg  
https://output.circle-artifacts.com/output/job/bc6c4914-a5ab-4b17-82aa-d50a896215a7/artifacts/0/tmp/artifacts/mattermost-desktop-5.3.0-develop.1-linux-arm64.tar.gz  
https://output.circle-artifacts.com/output/job/bc6c4914-a5ab-4b17-82aa-d50a896215a7/artifacts/0/tmp/artifacts/mattermost-desktop-5.3.0-develop.1-linux-x64.tar.gz  
https://output.circle-artifacts.com/output/job/584543de-9bda-4bae-aee0-2d052bd78927/artifacts/0/tmp/artifacts/mattermost-desktop-5.3.0-develop.1-win-arm64.zip  
https://output.circle-artifacts.com/output/job/584543de-9bda-4bae-aee0-2d052bd78927/artifacts/0/tmp/artifacts/mattermost-desktop-5.3.0-develop.1-win-ia32.zip  
https://output.circle-artifacts.com/output/job/584543de-9bda-4bae-aee0-2d052bd78927/artifacts/0/tmp/artifacts/mattermost-desktop-5.3.0-develop.1-win-x64.zip

mattermost-build commented:
Building app in separate branch.

mattermost-build commented:
Successfully building:  
https://circleci.com/gh/mattermost/desktop/27027  
https://circleci.com/gh/mattermost/desktop/27026  
https://circleci.com/gh/mattermost/desktop/27025

mattermost-build commented:
Artifact links:  
https://output.circle-artifacts.com/output/job/ae8b4c48-b387-46cb-a7b4-9af019a38de6/artifacts/0/tmp/artifacts/mattermost-desktop-5.3.0-develop.1-mac-m1.dmg  
https://output.circle-artifacts.com/output/job/ae8b4c48-b387-46cb-a7b4-9af019a38de6/artifacts/0/tmp/artifacts/mattermost-desktop-5.3.0-develop.1-mac-x64.dmg  
https://output.circle-artifacts.com/output/job/60b90dfe-97e8-4df6-9c4a-38812f23ecae/artifacts/0/tmp/artifacts/mattermost-desktop-5.3.0-develop.1-win-arm64.zip  
https://output.circle-artifacts.com/output/job/60b90dfe-97e8-4df6-9c4a-38812f23ecae/artifacts/0/tmp/artifacts/mattermost-desktop-5.3.0-develop.1-win-ia32.zip  
https://output.circle-artifacts.com/output/job/60b90dfe-97e8-4df6-9c4a-38812f23ecae/artifacts/0/tmp/artifacts/mattermost-desktop-5.3.0-develop.1-win-x64.zip  
https://output.circle-artifacts.com/output/job/5064faec-f50a-4f8b-8758-c24c5e5120a8/artifacts/0/tmp/artifacts/mattermost-desktop-5.3.0-develop.1-linux-arm64.tar.gz  
https://output.circle-artifacts.com/output/job/5064faec-f50a-4f8b-8758-c24c5e5120a8/artifacts/0/tmp/artifacts/mattermost-desktop-5.3.0-develop.1-linux-x64.tar.gz

jupenur commented on src/main/views/MattermostView.ts:
```diff
@@ -183,28 +177,65 @@ export class MattermostView extends EventEmitter {
+    }
+
+    private addCORSResponseHeader = (details: OnHeadersReceivedListenerDetails): HeadersReceivedResponse => {
+        if (!details.url.startsWith(this.tab.server.url.origin)) {
```
This check is broken and can be bypassed with XHR and credentials in URLs. If a user adds two servers, `mattermost.evil.com` and `community.mattermost.com`, the former can then do this:

```js
let xhr = new XMLHttpRequest();
xhr.open("GET", "https://mattermost.evil.com@community.mattermost.com/any/path");
xhr.send(null);
```

CORS headers will get added and the response will come through to the attacker. This needs to be fixed before merging.

devinbinnie replied:
I'll replace this with an exact origin check, that should fix this bug.

jupenur reviewed (changes requested):
SSO still allows evil tricks with navigation and that also needs to be fixed prior to release. And ideally we'd also fix the thing where after navigating to an origin that's not the current server URL, the origin then loads scripts from that URL.

But with the CORS thing it looks like this still deserves more attention before merging anyway. It's just too complex. I'll have to continue next week.

mattermost-build commented:
Building app in separate branch.

mattermost-build commented:
Successfully building:  
https://circleci.com/gh/mattermost/desktop/27047  
https://circleci.com/gh/mattermost/desktop/27046  
https://circleci.com/gh/mattermost/desktop/27045

mattermost-build commented:
Artifact links:  
https://output.circle-artifacts.com/output/job/db045bd6-1818-481d-8dce-dc0101d71b1e/artifacts/0/tmp/artifacts/mattermost-desktop-5.3.0-develop.1-mac-m1.dmg  
https://output.circle-artifacts.com/output/job/db045bd6-1818-481d-8dce-dc0101d71b1e/artifacts/0/tmp/artifacts/mattermost-desktop-5.3.0-develop.1-mac-x64.dmg  
https://output.circle-artifacts.com/output/job/d3397393-6de2-497b-88be-bfe7255ac3d4/artifacts/0/tmp/artifacts/mattermost-desktop-5.3.0-develop.1-win-arm64.zip  
https://output.circle-artifacts.com/output/job/d3397393-6de2-497b-88be-bfe7255ac3d4/artifacts/0/tmp/artifacts/mattermost-desktop-5.3.0-develop.1-win-ia32.zip  
https://output.circle-artifacts.com/output/job/d3397393-6de2-497b-88be-bfe7255ac3d4/artifacts/0/tmp/artifacts/mattermost-desktop-5.3.0-develop.1-win-x64.zip  
https://output.circle-artifacts.com/output/job/5011a9a5-fe97-4ba2-8069-243616bc56e3/artifacts/0/tmp/artifacts/mattermost-desktop-5.3.0-develop.1-linux-arm64.tar.gz  
https://output.circle-artifacts.com/output/job/5011a9a5-fe97-4ba2-8069-243616bc56e3/artifacts/0/tmp/artifacts/mattermost-desktop-5.3.0-develop.1-linux-x64.tar.gz

jupenur reviewed (approved):
Spotted another major issue that was introduced in this PR: https://github.com/mattermost/desktop/pull/2343

Cookies are not tied in any way to a specific origin, so if a webapp instance loads an external image, any cookies set on the server will be sent to that external host as well, and any cookies the external host sets get defined on the Mattermost instance.

Out-of-band we discussed with @devinbinnie regarding a strategy for reviewing all of this and I'm going to approve this now still knowing there are outstanding security issues. Later we'll do a proper security review once the feature branch is ready to be merged to master.

devinbinnie commented on src/main/views/MattermostView.ts:
```diff
@@ -113,28 +114,15 @@ export class MattermostView extends EventEmitter {
-            new RegExp(`file://(${this.tab.server.url.pathname})?/(api|static|plugins)/(.*)`, 'g'),
+            this.tab.server.url.host,
+            'mm-desktop:',
+            new RegExp(`^(${this.tab.server.url.pathname})?/(api|static|plugins)/(.*)`, 'g'),
```
Right, I was being overzealous here. Will remove the 'g'

devinbinnie commented:
> Spotted another major issue that was introduced in this PR: #2343
> 
> Cookies are not tied in any way to a specific origin, so if a webapp instance loads an external image, any cookies set on the server will be sent to that external host as well, and any cookies the external host sets get defined on the Mattermost instance.

Created a bug ticket here to deal with this issue after this PR is merged: https://mattermost.atlassian.net/browse/MM-49712
